staging/src/k8s.io/apiserver/plugin/pkg/audit/dynamic/dynamic.go

```diff
@@ -201,6 +201,7 @@ func (b *backend) Run(stopCh <-chan struct{}) error {
 // the primary stopChan to the current delegate map.
 func (b *backend) stopAllDelegates() {
 	b.delegateUpdateMutex.Lock()
+	defer b.delegateUpdateMutex.Unlock()
```

Comment:
See https://github.com/kubernetes/kubernetes/pull/70021#discussion_r232407338 for history. Leaving the mutex locked was intentional.

This should definitely have a comment explaining it though.

Reply:
my vague memory was correct, though the function comment should have accompanied the move, and leaving a hung lock is a minefield of a way to prevent additional delegate additions. A change to make delegate addition attempts no-op (and potentially log a warning) once stopCh/stopAllDelegates has been triggered would be much better

Reply:
> Leaving the mutex locked was intentional

This is not a good way of preventing additional changes (I assume that's what it's for?) because it results in hung goroutines. It needs to e.g. take the lock, set a "shutdown" flag, and release the lock, so that future lock-takers can read the shutdown flag and return an error rather than hanging.

Reply:
> This is not a good way of preventing additional changes (I assume that's what it's for?) because it results in hung goroutines.

Exactly

Reply:
Hi all, thanks a lot for catching my mistake! Otherwise I would submit a wrong PR.
I assume we need to let update functions return before acquiring the Mutex. My plan is to add a new field in `backend`: `b.delegateDisableCh`, an unbuffered channel of type `struct{}`. 
I will close this channel in `stopAllDelegates()`, and put a select in update functions before they try to acquire the Mutex. 
The select will do nothing if the channel is not closed, but will call `blog.Errorf("blabla...delegate update is disabled...")` if this channel is closed.
Do you think this patch is OK?

Reply:
I don't think you need a channel, just a bool. Basically, everywhere the mutex is locked, add this check after:

```
if b.stopped {
  return
}
```

And in `stopAllDelegates`:
```
b.delegateUpdateMutex.Lock()
defer b.delegateUpdateMutex.Unlock()
if b.stopped {
  return
}
b.stopped = true
```

I can't quite remember the reason that stopAllDelagates is separate from shutdown. I think it's captured in the comments on the original PR.

Reply:
Sure, this is less expensive than select, I suppose. Then do you think I need to call ‘klog.Errorf’ to record ‘stopped == true’ in update functions?